src/main/scala/encry/view/history/HistoryImpl.scala

```diff
@@ -0,0 +1,205 @@
+package encry.view.history
+
+import java.io.File
+import com.typesafe.scalalogging.StrictLogging
+import encry.consensus.History.ProgressInfo
+import encry.modifiers.history._
+import encry.settings._
+import encry.storage.VersionalStorage
+import encry.storage.VersionalStorage.{StorageKey, StorageValue, StorageVersion}
+import encry.storage.iodb.versionalIODB.IODBHistoryWrapper
+import encry.storage.levelDb.versionalLevelDB.{LevelDbFactory, VLDBWrapper, VersionalLevelDBCompanion}
+import encry.utils.NetworkTimeProvider
+import encry.view.history.storage.HistoryStorage
+import io.iohk.iodb.{ByteArrayWrapper, LSMStore}
+import org.encryfoundation.common.modifiers.PersistentModifier
+import org.encryfoundation.common.modifiers.history.{Block, Header, Payload}
+import org.encryfoundation.common.utils.Algos
+import org.encryfoundation.common.utils.TaggedTypes.ModifierId
+import org.iq80.leveldb.Options
+import cats.syntax.either._
+
+/**
+  * History implementation. It is processing persistent modifiers generated locally or received from the network.
+  **/
+trait HistoryImpl extends HistoryModifiersValidator with HistoryModifiersProcessors with AutoCloseable {
+
+  def isFullChainSynced: Boolean = getBestHeaderId
+    .exists(bestHeaderId => getBestBlockId.exists(bId => ByteArrayWrapper(bId) == ByteArrayWrapper(bestHeaderId)))
+
+  /** Appends modifier to the history if it is applicable. */
+  def append(modifier: PersistentModifier): Either[Throwable, (HistoryImpl, ProgressInfo)] = {
+    logger.info(s"Trying to append modifier ${Algos.encode(modifier.id)} of type ${modifier.modifierTypeId} to history")
+    Either.catchNonFatal(modifier match {
+      case header: Header => (this, process(header))
+      case payload: Payload => (this, process(payload))
```

Comment:
You can remove matching from here, because it would be performed in `process` method